web/tests/functional/export/test_export.py

```diff
@@ -0,0 +1,116 @@
+
+
+def get_user_comments(comments):
+    """
+    Separate comments and returns a tuple of user comments and system.
+    """
+    user_comments = [c for c in comments
+                     if c.kind == CommentKind.USER]
+
+    return user_comments
+
+
+class TestExport(unittest.TestCase):
+
+    _ccClient = None
+
+    def setUp(self):
+
+        # TEST_WORKSPACE is automatically set by test package __init__.py .
+        test_workspace = os.environ['TEST_WORKSPACE']
+
+        test_class = self.__class__.__name__
+        print('Running ' + test_class + ' tests in ' + test_workspace)
+
+        # Get the test project configuration from the prepared test workspace.
+        self._testproject_data = env.setup_test_proj_cfg(test_workspace)
+        self.assertIsNotNone(self._testproject_data)
+
+        # Setup a viewer client to test viewer API calls.
+        self._cc_client = env.setup_viewer_client(test_workspace)
+        self.assertIsNotNone(self._cc_client)
+
+        # Get the CodeChecker cmd if needed for the tests.
+        self._codechecker_cmd = env.codechecker_cmd()
+
+        # Get the run names which belong to this test.
+        run_names = env.get_run_names(test_workspace)
+
+        runs = self._cc_client.getRunData(None, None, 0, None)
+        self.test_runs = [run for run in runs if run.name in run_names]
+
+    def test_export(self):
+        """
+        Test exporting feature
+        """
+        runid = self.test_runs[0].runId
+        run_filter = RunFilter([runid])
+        logging.debug('Get all run results from the db for runid: ' +
+                      str(runid))
+
+        run_results = get_all_run_results(self._cc_client, runid)
+        self.assertIsNotNone(run_results)
+        self.assertNotEqual(len(run_results), 0)
+
+        bug = run_results[0]
+
+        # There are no comments available for the first bug
+        comments = self._cc_client.getComments(bug.reportId)
+        self.assertEqual(len(comments), 0)
+
+        comment1 = CommentData(author='anybody', message='First msg')
+        success = self._cc_client.addComment(bug.reportId, comment1)
+        self.assertTrue(success)
+        logging.debug('Bug commented successfully')
+
+        # Try to add another new comment for the first bug
+        comment2 = CommentData(author='anybody', message='Second msg')
+        success = self._cc_client.addComment(bug.reportId, comment2)
+        self.assertTrue(success)
+        logging.debug('Bug commented successfully')
+
+        # Change review status
+        review_comment = 'This is really a bug'
+        status = ReviewStatus.CONFIRMED
+        success = self._cc_client.changeReviewStatus(
+            bug.reportId, status, review_comment)
+
+        self.assertTrue(success)
+        logging.debug('Bug review status changed successfully')
+
+        # Export the comments and reviews
+        exported = self._cc_client.exportData(run_filter)
```

Comment:
Here you test only the export feature. Can we test the import feature too somehow?

For example at the end of this test case we remove some comments and then we can import it again.

Reply:
Hmm. I think this can be done, but then would you suggest changing the same of the root directory of this folder. Cause as of now, it is just named as `importer`.

Another approach would be making a separate folder for the importer and doing all the steps done for the importer and actually importing actual json file, but IMO that will be much harder. 

I think we would go with the first approach itselt, what do you think @csordasmarton ??

Reply:
For now I think it will be good to do the following:
- Rename the root folder to `export_import`.
- Rename the test file to `test_export_import.py`.
- Rename the test case from `test_export` to `test_export_import`.
- What you did in this test case is good. The next thing we should do in this test case is to export the data to a json file similar to this: https://github.com/Ericsson/codechecker/blob/dd194907919f020b2b825bf1f66f2b544e855db4/web/tests/functional/diff_local/test_diff_local.py#L74-L80. So you have to call the CodeChecker cmd export command and print the output to a temporary json file. For example you can create a temp file like this: https://github.com/Ericsson/codechecker/blob/dd194907919f020b2b825bf1f66f2b544e855db4/web/tests/functional/cmdline/test_cmdline.py#L268
- Next remove the first comment by using the `removeComment` API function.
- Update the message of the second comment by using the `updateComment` API function.
- Call the CodeChecker cmd import function with the temporary json file created above.
- Check whether the data is imported to the server. You can use the `exportData` method similar what you did before and see if it contains results from the import set or you can use the `getComments` API function and check the results.

Reply:
@csordasmarton writing it to a json file would require the formatting for again converting the output from json into exportData format(https://github.com/jay24rajput/codechecker/blob/e47a86d23cb9ca2454cfd9c561c5a543b07330ff/web/client/codechecker_client/cmd_line_client.py#L1681) whereas here we already have an exportData object exported(https://github.com/jay24rajput/codechecker/blob/e47a86d23cb9ca2454cfd9c561c5a543b07330ff/web/tests/functional/export/test_export.py#L105) from which it is easy to delete/modify comments and review status and again create a new exportData object directly. I think for the testing purpose writing it to a json format and again importing it would be kind of redundant since we are testing the `importdata` method which only takes an exportData object as an input

Reply:
Yes, you are right. Here we have the data but I recommend to use the commands (CodeChecker cmd export/import) instead of calling API functions. The test will more robust, because on the client side commands we do a couple of extra things. And most of the user will use these commands instead of the API functions. And with this we would test whether the conversion to json works properly, it is a valid json file, we can import this json file to the server etc.

Reply:
@csordasmarton In the following code if executed for the export_import, it shows error in the subprocess.check command.
https://github.com/jay24rajput/codechecker/blob/cmd_importer/web/tests/functional/export_import/test_export_import.py

I am attaching the stack trace of the same:
```
ERROR: Test exporting and importing feature
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jay/codechecker/web/tests/functional/export_import/test_export_import.py", line 122, in test_export
    out_json = subprocess.check_output(new_export_command, encoding="utf-8", errors="ignore")
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/jay/codechecker/build/CodeChecker/bin/CodeChecker', 'cmd', 'export', '-n', 'test_files_5f78c3e2b3c34bfcbcade4c2eb1af76e']' returned non-zero exit status 1.
-------------------- >> begin captured stdout << ---------------------
Running TestExport tests in /home/jay/codechecker/build/workspace/export_import-3_0qq4zf
Could not load session for session getter because 'localhost:45537'
Setup viewer client: http://localhost:45537/export_import/v6.37/CodeCheckerService
['/home/jay/codechecker/build/CodeChecker/bin/CodeChecker', 'cmd', 'export', '-n', 'test_files_5f78c3e2b3c34bfcbcade4c2eb1af76e']
```

May you know why is error is occurring? Cause the syntax for running the command is:
```sh
CodeChecker cmd export -n <run_name>
```
May you know why this error? If we comment out the lines suggested by you the test runs successfully

Reply:
Sorry, my bad. I gave you a wrong example. In our case you have to specify the `--url` option to this command too:
https://github.com/Ericsson/codechecker/blob/9d23b7ba6d4c149a6fabf21579f08fc60ef4df1d/web/tests/functional/cmdline/test_cmdline.py#L251

In the `setUp` function you can use this:
https://github.com/Ericsson/codechecker/blob/9d23b7ba6d4c149a6fabf21579f08fc60ef4df1d/web/tests/functional/cmdline/test_cmdline.py#L50-L52

Reply:
@csordasmarton Yes that seems to work! Thanks a lot! For the import command I am planning to do something like:
```python
 with tempfile.NamedTemporaryFile() as component_f:
            component_f.write(out_json.encode('utf-8'))
           # Will delete and update the comments here in codechecker server

            new_import_command = [self._codechecker_cmd, 'cmd', 'import',
                                  '-i', component_f.name,
                                  '--url', str(self.server_url)]
            print(new_import_command)
            output = subprocess.check_output(
                new_export_command, encoding="utf-8", errors="ignore")
```

Does this method seem right to you? Or you have any other method of doing the same?

Reply:
It looks good except that in the last line you are calling the export command (`new_export_command`) and not the import (`new_import_command`) :blush:

Reply:
Sadly even this is returning the following error:
```
ERROR: Test exporting and importing feature
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jay/codechecker/web/tests/functional/export_import/test_export_import.py", line 145, in test_export
    output = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/jay/codechecker/build/CodeChecker/bin/CodeChecker', 'cmd', 'import', '-i', '/tmp/tmpa5nvjjzi', '--url', 'localhost:43645/export_import']' returned non-zero exit status 
```

Reply:
I think the problem will be that you have to `flush` the output after calling the `write` method:
```py
component_f.write(out_json.encode('utf-8'))
component_f.flush()
component_f.seek(0)
````

Also you can catch these error with a simple try/except blocks:
```py
try:
    output = subprocess.check_output(
        new_import_command, encoding="utf-8", errors="ignore")
except subprocess.CalledProcessError as e:
    print(e.output)
```

Also if you have such problem, please don't forget to update your branch because it will be easier for me to test it.